### DISCUSSION TITLE
Is any documentation to write custom application handler (module) ?

### DISCUSSION THREAD

**Discussion by tregubovav-dev** [nginx/unit#1102] [Support]
Hello Nginx Unit development team.

Nginx Unit great platform for serverless applications. However, it looks as good platform for native programming languages which hard to port to Nginx Server due to internal API complexity.
I'm looking for any documentation to build additional module for currently unsupported frameworks.
Looking into Python/Perl/Ruby modules I see I can register custom module dynamically. However, handling module configuration, http requests and returning http responses by app require researching source code deeply. 

I would appreciate it if you can point me any design guidelines and/or module API descriptions.

**Comment by ac000** [comment]
> Hello Nginx Unit development team.

Hi,

> Nginx Unit great platform for serverless applications. However, it looks as good platform for native programming languages which hard to port to Nginx Server due to internal API complexity. I'm looking for any documentation to build additional module for currently unsupported 

No documentation as such...

> frameworks. Looking into Python/Perl/Ruby modules I see I can register custom module dynamically. However, handling module configuration, http requests and returning http responses by app require researching source code deeply.

A lot of the heavy lifting is taken care of for you, your main tasks of concern are handling the request and sending the response.

> I would appreciate it if you can point me any design guidelines and/or module API descriptions.

[Here](https://git.digital-domain.net/unit.git/tree/src/skel/nxt_skel.c?h=skel_lm) is an example skeleton module.

See this [commit](https://git.digital-domain.net/unit.git/commit/?h=skel_lm&id=19c9234a64b4ee5d2622c874104f31c7a0987f25) for how it hooks into the Unit core.

That's enough to build the module and load it.

For a real module, our [WebAssembly module](https://github.com/nginx/unit/tree/master/src/wasm) is perhaps the clearest to see how it works.

nxt_wasm.[ch] are the main interface to Unit, handling module loading/setup and requests/responses.

For the configuration, this [commit](https://github.com/nginx/unit/commit/e99854afdf555678df5af149e53659c5b30a2f6d) shows how it's hooked up to the Unit configuration infrastructure.

Then [here](https://github.com/nginx/unit/blob/990fbe7010526bb97f2d414db866050ef5e8f244/src/wasm/nxt_wasm.c#L251) we get a pointer to the config data setup in that last commit.

Then a few lines down we can simply access them as in this case they are largely const char *'s. You can see the structure member names match up with what's in the above commit.

[This](https://github.com/nginx/unit/blob/master/src/nxt_unit.h) also contains a description of some of the API used.

**Comment by tregubovav-dev** [comment]
Thank you Andrew for your quick response.
I appreciate for your suggestion to real code. This could simplify understanding the architecture and API.
